plasmapy/utils/decorators/tests/test_checks.py

```diff
@@ -10,10 +11,939 @@
+
+            if _results[3] is None:
+                assert _results[3] is case['output'][3]
+                assert cu._check_unit(arg, arg_name, arg_checks) is None
+            else:
+                assert isinstance(_results[3], case['output'][3])
+                with pytest.raises(case['output'][3]):
+                    cu._check_unit(arg, arg_name, arg_checks)
+
+    def test_cu_called_as_decorator(self):
+        """
+        Test behavior of `CheckUnits.__call__` (i.e. used as a decorator).
+        """
+        # setup test cases
+        # 'setup' = arguments for `CheckUnits` and wrapped function
+        # 'output' = expected return from wrapped function
+        # 'raises' = if an Exception is expected to be raised
+        # 'warns' = if a warning is expected to be issued
+        #
+        _cases = [
+            # clean execution
+            {'setup': {'function': self.foo_no_anno,
+                       'args': (2 * u.cm, 3 * u.cm),
+                       'kwargs': {},
+                       'checks': {'x': u.cm,
+                                  'y': u.cm,
+                                  'checks_on_return': u.cm}},
+             'output': 5 * u.cm,
+             },
+
+            # argument fails checks
+            {'setup': {'function': self.foo_no_anno,
+                       'args': (2 * u.cm, 3 * u.cm),
+                       'kwargs': {},
+                       'checks': {'x': u.g,
+                                  'y': u.cm,
+                                  'checks_on_return': u.cm}},
+             'raises': u.UnitTypeError,
+             },
+
+            # return fails checks
+            {'setup': {'function': self.foo_no_anno,
+                       'args': (2 * u.cm, 3 * u.cm),
+                       'kwargs': {},
+                       'checks': {'x': u.cm,
+                                  'y': u.cm,
+                                  'checks_on_return': u.km}},
+             'raises': u.UnitTypeError,
+             },
+        ]
+
+        # test
+        for case in _cases:
+            wfoo = CheckUnits(**case['setup']['checks'])(case['setup']['function'])
+
+            args = case['setup']['args']
+            kwargs = case['setup']['kwargs']
+
+            if 'raises' in case:
+                with pytest.raises(case['raises']):
+                    wfoo(*args, **kwargs)
+            else:
+                assert wfoo(*args, **kwargs) == case['output']
+
+    def test_cu_preserves_signature(self):
+        """Test `CheckValues` preserves signature of wrapped function."""
+        # I'd like to directly dest the @preserve_signature is used (??)
+
+        wfoo = CheckUnits()(self.foo_no_anno)
+        assert hasattr(wfoo, '__signature__')
+        assert wfoo.__signature__ == inspect.signature(self.foo_no_anno)
+
+    @mock.patch(CheckUnits.__module__ + '.' + CheckUnits.__qualname__,
+                side_effect=CheckUnits, autospec=True)
```

Comment:
What's the point of patching this?

Reply:
This patch creates the `mock_cu_class` Mock object, which is used to ensure the `check_units` decorator calls the `CheckUnits` class with the correct arguments (see lines [541-549](https://github.com/rocco8773/PlasmaPy/blob/576e49bb3ef2d909833449cc9be918a28dbbb276/plasmapy/utils/decorators/tests/test_checks.py#L541-L549) below).  This then allows me to do 99% of my tests on `CheckUnits` and not have to duplicate them on `check_units`.